src/lib/threadpool.c

```diff
@@ -0,0 +1,529 @@
+	uv_thread_t  thread;    /* Pool's worker thread */
+	struct targs arg;
+};
+
+struct pool_impl {
+	uv_mutex_t     mutex;           /* Input queue, planner_sm,
+                                           worker and planner threads lock */
+	uint32_t       threads_nr;
+	pool_thread_t *threads;
+
+	queue          outq;            /* Output queue used by libuv part */
+	uv_mutex_t     outq_mutex;      /* Output queue lock */
+	uv_async_t     outq_async;      /* Signalled when output queue is not
+                                           empty and libuv loop has to process
+                                           items from it */
+	uint64_t       active_ws;       /* Number of all work items in flight,
+					   accessed from the main thread only */
+
+	queue          ordered;         /* Queue of WT_ORD{N} items */
+	queue          unordered;       /* Queue of WT_UNORD items */
+	struct sm      planner_sm;      /* State machine of the scheduler */
+	uv_cond_t      planner_cond;
+	uv_thread_t    planner_thread;  /* Scheduler's thread */
+
+	uint32_t       ord_in_flight;   /* Number of WT_ORD{N} in flight */
+	bool           exiting;         /* True when the pool is being stopped */
+	enum pool_work_type             /* Type of the previous work item, */
+                       ord_prev;        /* used in WT_ORD{N} ivariants */
+	uint32_t       qos;             /* QoS token */
+	uint32_t       qos_prio;        /* QoS prio */
+};
+
+static inline bool has_active_ws(pool_t *pool)
+{
+	return pool->pi->active_ws > 0;
+}
+
+static inline void w_register(pool_t *pool, pool_work_t *w)
+{
+	if (w->type != WT_BAR)
+		pool->pi->active_ws++;
+}
+
+static inline void w_unregister(pool_t *pool, pool_work_t *)
+{
+	PRE(has_active_ws(pool));
+	pool->pi->active_ws--;
+}
+
+static bool empty(const queue *q)
+{
+	return QUEUE__IS_EMPTY(q);
+}
+
+static queue *head(const queue *q)
+{
+	return QUEUE__HEAD(q);
+}
+
+static void push(queue *to, queue *what)
+{
+	QUEUE__INSERT_TAIL(to, what);
+}
+
+static queue *pop(queue *from)
+{
+	queue *q = QUEUE__HEAD(from);
+	PRE(q != NULL);
+	QUEUE__REMOVE(q);
+	QUEUE__INIT(q);
+	return q;
+}
+
+static queue *qos_pop(pool_impl_t *pi, queue *first, queue *second)
+{
+	PRE(!empty(first) || !empty(second));
+
+	if (empty(first))
+		return pop(second);
+	else if (empty(second))
+		return pop(first);
+
+	return pop(pi->qos++ % pi->qos_prio ? first : second);
+}
+
+static pool_work_t *q_to_w(const queue *q)
+{
+	return QUEUE__DATA(q, pool_work_t, link);
+}
+
+static enum pool_work_type q_type(const queue *q)
+{
+	return q_to_w(q)->type;
+}
+
+static uint32_t q_tid(const queue *q)
+{
+	return q_to_w(q)->thread_id;
+}
+
+static bool planner_invariant(const struct sm *m, int prev_state)
+{
+	pool_impl_t *pi = container_of(m, pool_impl_t, planner_sm);
+	queue *o = &pi->ordered;
+	queue *u = &pi->unordered;
+
+	return ERGO(sm_state(m) == PS_NOTHING, empty(o) && empty(u)) &&
```

Comment:
I think `planner_invariant` returns `true` when `sm_state(m)` is not one of the enum values, because the first argument to `ERGO` will always be `false`, that's probably unwanted.

Reply:
it's impossible to reach such condition due to the following reasons:
sm_init() will move the state to initial state and even if the user will pass incorrect state to sm_move(), PRE(m->conf[sm_state(m)].allowed & BITS(next_state)) will catch this.